README.md


## Basic Deploy
run web server:
```
	./app.py
```

## Test
To learn better run
```
python tests/test_simple.py
```

------------


## Tutorial

- in the first step you should define a workflow to server 
   to define new workflow call below address using standard HTTP call:
 
```
http POST http://server_address/wf/{workflow name:string}/{workflow version:number}
```

- content must be a json which defines our workflow. for example:

```
http POST http://server_address/wf/my_test_workflow/1
```

- json which sends to server must have below schema:
```
{ 
   	"name" : "workflow name",
	"steps":[
				{type= "taskcall", taskname="taskname to call in this step", params = [parameters to pass to task]}
				...
```

or
				
```
{
	type="eval",
	params = [ ''' string to evaluate ''']
}
				...	
```

### Full example:

```
http POST http://server_address/wf/my_test_workflow/1
```

```

{
        "name":"my_test_workflow",
            "steps":[
                        {
                            "type":"taskcall",
                            "taskname":"test_task_add_two_number",
                            "params":[12,13,"c"]
                        }
                    ]
}

```


this sample defines a simple workflow that add two number. ``test_task_add_two_number`` is a task which defined in our workflow server as below:

```
def test_task_add_two_number(a,b,
					result_var_name):
    	def l(obj,eng):
        	obj[result_var_name]=a+b
    	return l
```

as you see, our function has 3 parameter that pass to it using params section of task call in our json. all tasks must defined to our workflow server

To run this workflow :

```
POST http://server_address/run/{workflow name}/{workflow version}/{customer id}
```
And Body: 

json that initialize workflow variables
	``customer id`` is an identifcation number which specifies this ``wf(workflow)`` instance

in the case of VAS service it is user mobile number that this wf runs for.
to run pervious workflow:
```
http POST http://server_address/run/my_test_workflow/1/242233
```
``Content-Type``: ``application/json``
	
To get variables from workflow after its run we should defines which
variables we need, see below:

```
http POST http://server_address/run/my_test_workflow/1/123232
```
```
	{
		"__rtn_vars__":["c"]
	}
```
	
above says to workflow engine (WE) runs ``my_test_workflow`` for customer id ``123232`` and returns variable ``c``.	as you know c must be ``25`` (12+13)

to improve our understanding see this example:
```
{
	"name":"simple_login"
   "steps":[
      {
	  	 "taskname":"IF_ELSE",
         "type":"taskcall",
         "params":[
				{
				   "type":"eval",
				   "params":[
					  " lambda obj,eng: obj['username']=='gholi' and obj['password']=='gholi' "
				   ]
				},
					{
					   "taskname":"task_exec",
					   "type":"taskcall"
						"params":[
						  "
						  def l(obj,eng):
						  	obj['rtn']= 'hello dear '+obj['username']
							obj['user_login']=True
							",
						  "l"
					   	],
					},
					{
					   "taskname":"task_exec",
					   "type":"taskcall"
					   "params":[
						  "
						  def l(obj,eng):
						  	obj['rtn']='failed to Login'\n    obj['user_login']=False
							",
						  "l"
					   ],
					}
         ],
         
      }
   ],
   
}
```

Explanation:
- this example created by 1 task, it is ``IF_ELSE`` task, ``IF_ELSE`` task has 3 arguments
- first is condition which is a lambda expression to determine ``True`` of ``False``.
- second parameter is list of tasks if condition was ``True``.
 - Thrid parameter is list of tasks if condition was ``False``.

- As you see first parameter of ``IF_ELSE`` task is:

```
 {	
	"type":"eval",
	"params":[
			" lambda obj,eng: obj['username']=='gholi' and obj['password']=='gholi' "
		      ]
}
```
which defines an eval task that compiles to :
```
lambda obj,eng: obj['username']=='gholi' and obj['password']=='gholi'
```
that compare username and password to gholi
and if both of them are gholi returns True else returns ``False``.
	
- the second parameter is (True Branch of IF_ELSE [can be list of tasks]):

```	
{
	"taskname":"task_exec",
	"type":"taskcall",
	"params":[
		'''
		def l(obj,eng):
			obj['rtn']= 'hello dear '
				+obj['username']
			obj['user_login']=True
		''',
		"l"
	],
}
```

it is a task_exec which exec string in server, task_exec has two parameters:
```
task_exec(code_string,function_name_string)
```

code_string can be full Python Module but at final should define a function that has two parameter ``(obj,eng)`` that returns from it.

as you see this Task sets rtn variable of WF to "hello dear {username}" and 
	sets user_login variable to ```True```

to run this WF we call run method using:
```	
POST http://server_address/wf/run/simple_login/1/12344
```

```
	{
		'__rtn_vars__':['rtn','user_login'],
		'username':'gholi',
		'password':'jafar'
	}
```

above call sets username and password variable of workflow to 'gholi' and 'jafar' and 	says to workflow returns rtn,user_login variables from workflow.

----

Sometimes we need call external web-api and interact with them to do this we do below steps:
	1. define web url and its type to server using below call:
```
POST http://server_address/wf/define/task
```

```
{
	"task_name":"task name which use in WF to specifies this web task",
	"task_type":"can be sync or async"
	"url":"url of server method"
}
```

sync tasks wait to get response from server and workflow waits for the result
async tasks halts workflow and wait to call from web server to callbacked

for example:
```
POST http://server_address/wf/define/task
```
```
{
		"task_name":"reg_user",
		"task_type":"sync",
		"url":"http://127.0.0.1:8282/mockservice/user"
}
```

	2. To use in our wf we should define a task as below:
```
{
	"type":"taskcall",
	"taskname":"task_dynamic_task",
    params=[
			"web task name",
			"prefix of return result",
				{
                   web_call_param1:{type='var or value',var_name='variable name' or value = "direct value"},
					 ....
                 }]
}
```

For example:
``` 
{
	type='taskcall',
	taskname='task_dynamic_task',
    params=[
			'reg_user',
			'ru',
			{
               username:{type:'var',var_name:'username'},
               password:{type:'var',var_name:'password'},
               req_date:{type:'value',value:13931413},
               from_service_name:{type:'value',value:'simple test to farsheed!'}
            }
		]
}
```

in this sample WF calls ``reg_user`` web task which defined before and add ``ru`` prefix in the head of its returns json

for example if this web call returns json as below:
```
{userid:12,status:'ok'}
```

the wf add re as prefix to them and add them to wf variables,	``re_userid=12`` and ``re_status='ok'`` added to wf variables

last parameter is a dictionary which pass as parameter to web call
``username`` variable from workflow pass to web call as ``username`` in json body
``password`` variable from wf pass to web-call as ``password`` in json body
``req_date`` as direct value ``13931413`` pass to web-call as ``req_date`` 	and so on.
	
	
	
	
   
   


CHANGELOG.txt
11 OCT 2015:
    Web_server.py rename to app.py
    tasks moved to tasks folder



tests/msd_test2.py
from random import randrange

__author__ = 'msd'
from sys import stdin
from msdworkflow import msd_workflow
import workflow.patterns as wp


import os.path
import os


#---------- TASK DEFINITIONS
def send_sms(message):
    def l(obj, eng):
        print message, " sent to ", eng.getVar('usermob')
    return l


def sleep_for_seconds(seconds):
    def l(obj, eng):
        print "[SYSMSG]run this process after ", seconds, "(s),again and fill input fields by these values:"
        print "User Mobile Number:", eng.getVar('usermob')
        print "Workflow Name:", eng._process_name
        eng.haltProcessing()
    return l


def select_random_english_word(obj, eng):
    words = dict(cat='pishih', water='Su', nothing='Hava , Hech',
                 father='dada', mother='ana')
    m = randrange(0, len(words.keys()))
    word = words.keys()[m]
    mean = words[word]
    obj['word'] = word
    obj['mean'] = mean


def send_format_message(formatstr, variablenamesarray):
    def l(obj, eng):
        vars = dict()
        for i in variablenamesarray:
            vars[i] = obj[i]
        print formatstr.format(**vars), " sent to ", eng.getVar('usermob')
    return l


def get_user_guess(obj, eng):
    obj['user_guess'] = stdin.readline().strip()
#---------------------------------------

our_workflows = {'eng_lang': [
    send_sms('Hello Dear Friend'),
    send_sms('This is Only Test'),
    sleep_for_seconds(12),
    send_sms('hi, im waked up ...'),
    send_sms('lets go to learn english words'),
    select_random_english_word,
    send_format_message(
        'do u know meaning of {word}? it is {mean}', ['word', 'mean']),
    sleep_for_seconds(3600),
    send_format_message('Oh! What is {word} meaning?', ['word']),
    get_user_guess,
    wp.IF_ELSE(lambda obj, eng:obj['user_guess'] == obj['mean'],
               send_sms('congratulation'),
               send_sms('it was wrong. finish')
               )

]}


def get_workflow_by_name(wf_name):
    return our_workflows[wf_name]


def main():
    print "Enter User Mobile Number:"
    usermobile = stdin.readline().strip()
    print "Enter Flow Name (if dont know any thing pls enter eng_lang):"
    wf_name = stdin.readline().strip()
    file_name = wf_name + '_' + usermobile
    if not os.path.exists(file_name):
        # user in progress continue its processing
        run_new_wf(wf_name, usermobile)
    else:
        #user is new
        continue_wf(wf_name, usermobile)


def continue_wf(wf_name, usermob):
    file_name = wf_name + '_' + usermob
    f = open(file_name, 'rb')
    file_content = f.read()
    f.close()
    wf = msd_workflow.deserialize_workflow(file_content, get_workflow_by_name)
    is_finished = wf.continue_process()
    after_run_wf(wf, is_finished)


def run_new_wf(wf_name, usermob):
    wf = msd_workflow(wf_name, get_workflow_by_name)
    wf.setVar('usermob', usermob)
    is_finished = wf.start({'word': None, 'mean': None})
    after_run_wf(wf, is_finished)


def after_run_wf(wf, is_finished):
    filename = wf._process_name + '_' + wf.getVar('usermob')
    if not is_finished:
        # save wf and to further run
        tmp = wf.serialize_workflow()
        f = open(filename, 'wb')
        f.write(tmp)
        f.close()
    else:
        # delete file
        os.remove(filename)


if __name__ == '__main__':
    main()


tests/test_simple.py
__author__ = 'msd'
wf_base_address = 'http://127.0.0.1:8181/wf'
from json import dumps, loads

import requests
# wf = dict(name='eng_lang',
#           steps = [
#               dict(type='taskcall',taskname='send_sms',params=['Hello Dear Friend']),
#               dict(type='taskcall',taskname='send_sms',params=['This is Only Test']),
#               dict(type='taskcall',taskname='sleep_for_seconds',params=[12]),
#               dict(type="taskcall",taskname="send_sms",params=['hi, im waked up ...']),
#               dict(type='taskcall',taskname="send_sms",params=['lets go to learn english words']),
#               dict(type='taskcall',taskname='task_select_random_english_word',params=[]),
#               dict(type='taskcall',taskname='send_format_message',params=['do u know meaning of {word}? it is {mean}', ['word', 'mean']]),
#               dict(type='taskcall',taskname='sleep_for_seconds',params=[26]),
#               dict(type='taskcall',taskname='send_format_message',params=['Oh! What is {word} meaning?', ['word']]),
#               dict(type='taskcall',taskname='task_get_user_guess',params=[]),
#               dict(type='taskcall',taskname='IF_ELSE',params=[dict(type='eval',params=["""lambda obj, eng: obj['user_guess'] == obj['mean']"""]),
#                                                               dict(type='taskcall',taskname='send_sms',params=['congratulation']),
#                                                               dict(type='taskcall',taskname='send_sms',params=['it was wrong. finish'])])
#           ])


def run_wf1():
    res = requests.post(
        "{baseurl}/run/{wfname}/{wfver}/{custid}".format(
            baseurl=wf_base_address, wfname='simple_add_two_number', wfver=1, custid=123),
        data=dumps({'__rtn_vars__': ['c']}),
        headers={'Content-type': 'application/json'}
    )
    assert isinstance(res, requests.Response)
    if res.status_code > 299:
        raise 'Problem in run workflow wf1 status code', res.status_code
    print res.content


def def_wf1():
    simp_d = dict(name='simple_add_two_number',
                  steps=[
                      dict(type='taskcall', taskname='test_task_add_two_number', params=[
                           12, 13, 'c'])
                  ])
    # define wf to server
    a = requests.post(
        "{baseurl}/{wfname}/{wfver}".format(
            baseurl=wf_base_address, wfname='simple_add_two_number', wfver=1),
        data=dumps(simp_d),
        headers={'Content-type': 'application/json'}
    )
    assert isinstance(a, requests.Response)
    if a.status_code > 299:
        raise 'wf server response is,', a.status_code
    print "workflow defined to wf server"


def def_wf2():
    w = dict(name='simple_login',
             steps=[
                 dict(type='taskcall', taskname='IF_ELSE', params=[
                     dict(type='eval', params=[
                          ''' lambda obj,eng: obj['username']=='gholi' and obj['password']=='gholi' ''']),
                     # if username and password is correct:
                     dict(type='taskcall', taskname='task_exec', params=['''
def l(obj,eng):
    obj['rtn']= 'hello dear '+obj['username']
    obj['user_login']=True
                     ''', 'l']),
                     # if username and password is not correct
                     dict(type='taskcall', taskname='task_exec', params=['''
def l(obj,eng):
    obj['rtn']='failed to Login'
    obj['user_login']=False
                     ''', 'l'])
                 ])
             ])
    # define wf to server
    a = requests.post(
        "{baseurl}/{wfname}/{wfver}".format(
            baseurl=wf_base_address, wfname='simple_login', wfver=1),
        data=dumps(w),
        headers={'Content-type': 'application/json'}
    )
    assert isinstance(a, requests.Response)
    if a.status_code > 299:
        raise 'wf server response is,', a.status_code
    print "workflow defined to wf server"


def run_wf2(username, password):
    res = requests.post(
        "{baseurl}/run/{wfname}/{wfver}/{custid}".format(
            baseurl=wf_base_address, wfname='simple_login', wfver=1, custid=123),
        data=dumps({'__rtn_vars__': ['rtn', 'user_login'],
                    'username': username, 'password': password}),
        headers={'Content-type': 'application/json'}
    )
    assert isinstance(res, requests.Response)
    if res.status_code > 299:
        raise 'Problem in run workflow wf1 status code', res.status_code
    print res.content


def run_w3(username, password):
    res = requests.post(
        "{baseurl}/run/{wfname}/{wfver}/{custid}".format(
            baseurl=wf_base_address, wfname='register_user', wfver=1, custid=123),
        data=dumps({'__rtn_vars__': ['username', 'ru_status_code',
                                     'ru_body'], 'username': username, 'password': password}),
        headers={'Content-type': 'application/json'}
    )
    assert isinstance(res, requests.Response)
    if res.status_code > 299:
        raise 'Problem in run workflow wf1 status code', res.status_code
    print res.content


def def_w3():

    # define sync web call reg_user which add user to db
    taskdef = dict(task_name="reg_user", task_type="sync",
                   url="http://127.0.0.1:8282/mockservice/user")
    res = requests.post(
        "{baseurl}/task/define".format(baseurl=wf_base_address),
        data=dumps(taskdef),
        headers={'Content-type': 'application/json'}
    )
    assert isinstance(res, requests.Response)
    if res.status_code > 299:
        raise 'task definition problem ', res.status_code

    # define async webcall send_sms which sends sms asyncronsly
    taskdef = dict(task_name="send_sms_async", task_type="async",
                   url="http://127.0.0.1:8282/mockservice/async_sms")
    res = requests.post(
        "{baseurl}/task/define".format(baseurl=wf_base_address),
        data=dumps(taskdef),
        headers={'Content-type': 'application/json'}
    )
    assert isinstance(res, requests.Response)

    if res.status_code > 299:
        raise 'task definition problem ', res.status_code

    w = dict(
        name='register_user',
        steps=[
            dict(type='taskcall', taskname='task_dynamic_task',
                 params=['reg_user', 'ru', dict(
                     username=dict(type='var', var_name='username'),
                     password=dict(type='var', var_name='password'),
                     req_date=dict(type='value', value=13931413),
                     from_service_name=dict(
                         type='value', value='simple test to farsheed!')
                 )]),
            dict(type='taskcall', taskname='CHOICE', params=[
                dict(type='eval', params=[
                     '''lambda obj,eng: 'r'+str(obj['ru_status_code']) ''']),

                ['r200',
                 dict(type='taskcall', taskname='assign_value_to_var',
                      params=['STATUS', True]),
                 dict(type='taskcall', taskname='assign_var_to_var',
                      params=['user_id', 'ru_user_id']),
                 dict(type='taskcall', taskname='send_sms',
                      params=['You Are Joined Successfully']),
                 dict(type='taskcall', taskname='task_dynamic_task',
                      params=['send_sms_async', 'ssa',
                              dict(
                                  username=dict(
                                      type='var', var_name='username'),
                                  STATUS=dict(type='var', var_name='STATUS')
                              )]),
                 dict(type='taskcall', taskname='send_sms',
                      params=['FINISHED!'])
                 ], [
                    'r409',
                    dict(type='taskcall', taskname='assign_value_to_var',
                         params=['STATUS', False]),
                    dict(type='taskcall', taskname='send_sms',
                         params=['409 error happen'])

                ],
                ['r500', dict(type='taskcall', taskname='send_sms',
                              params=['500 error'])]

            ])

        ]
    )
    a = requests.post(
        "{baseurl}/{wfname}/{wfver}".format(
            baseurl=wf_base_address, wfname='register_user', wfver=1),
        data=dumps(w),
        headers={'Content-type': 'application/json'}
    )
    assert isinstance(a, requests.Response)
    if a.status_code > 299:
        print a.status_code
        print a.reason
        #raise 'wf server response is,',a.status_code
    print "workflow register_user defined to wf server"

if __name__ == '__main__':

    print "to run this test, sure that your workflow baseurl is :", wf_base_address
    raw_input("press any key to continue...")

    print 'define wf1 '
    def_wf1()
    print 'run wf1'
    run_wf1()
    print 'define simple login'
    def_wf2()
    print 'run simple logiin with username:gholi,password:gholi'
    run_wf2('gholi', 'gholi')
    print "run simple login with username:gholi1,password:1"
    run_wf2('gholi1', '1')
    raw_input(
        "to continue run mock_service.py to test web calls,press any key to continue ...")
    def_w3()
    run_w3('a', 'a')


tests/mock_service.py
__author__ = 'msd'

import random

from bottle import run, post, request, get, HTTPResponse
from json import dumps


@post('/mockservice/user')
def user():
    print request.json
    # this method add user to db
    # if user exists before returns 409
    # if user not exists and add successfully returns 200 and user id in json
    # if problem in add user returns 500
    if random.randint(0, 10) < 5:
        # 50% probiblity user exists before!!!!!
        return HTTPResponse(status=409, body=dumps(dict(message='try after 100s')), headers={'Content-type': 'application/json'})
    else:
        if random.randint(0, 10) < 2:
            # probeblity 20% error happend
            return HTTPResponse(status=500, body=dumps(dict(error='salam man error hastam')), headers={'Content-type': 'application/json'})
        else:
            # user added successfully
            return dict(user_id=random.randint(0, 100000))


@post('/mockservice/async_sms')
def asyn_sms():
    print 'received request:', request.json
    print "please manualy call this url after awhile:", request.json['__callback']


if __name__ == '__main__':
    run(host='127.0.0.1', port=8282)


app.py
#!/usr/bin/env python


import json

from bottle import run, post, request, get

from engine.msd_workflow_interface import run_workflow, compile_dict_to_workflow, run_workflow_, get_workflow_status
import engine.msd_wf_storage as storage


@post('/wf/<wfname>/<wf_version>/<custid>/status')
def get_status(wfname, wf_version, custid):
    return get_workflow_status(wfname, wf_version, custid)


@post('/wf/run/<wf_fullname>/<custid>')
def do_workflow_(wf_fullname, custid):
    rtn_var_names = []
    if '__rtn_vars__' in request.json:
        rtn_var_names = request.json.pop('__rtn_vars__')
    return run_workflow_(wf_fullname, custid, request.json, rtn_var_names)


@post('/wf/run/<wfname>/<wf_version>/<custid>')
def do_workflow(wfname, wf_version, custid):
    rtn_var_names = []
    if '__rtn_vars__' in request.json:
        rtn_var_names = request.json.pop('__rtn_vars__')
    return run_workflow(wfname, wf_version, custid, request.json, rtn_var_names)


@post('/wf/<wf_name>/<wf_version>')
def create_workflow(wf_name, wf_version):
    try:
        js = request.json['steps']
        compile_dict_to_workflow(js)
        storage.store(wf_name + "." + wf_version, json.dumps(js))
        return json.dumps(js)
    except TypeError as e:
        print e
        raise e


@post('/wf/task/define')
def define_dynamic_task():
    """
    define new task as below:
    {
        "task_name":"Task Name",
        "task_type":"sync or async",
        "url":"http://url.hostname.ext/path",

    }
    """
    storage.store('dynamic_task:{task_name}'.format(
        task_name=request.json['task_name']
    ), json.dumps(request.json))


@get('/wf/task/define/<taskname>')
def get_define_task_name(taskname):
    return storage.get('dynamic_task:{task_name}'.format(task_name=taskname))


@post('/wf/async_task_result/<wf_fullname>/<cust_id>/<prefix>')
def continue_async_dynamic_task(wf_fullname, cust_id, prefix):

    if hasattr(request, 'json') and request.json is not None:
        cor_dic = {"{RVP}_{KEY}".format(RVP=prefix, KEY=key): value for (
            key, value) in request.json.iteritems()}
    else:
        cor_dic = {}
    run_workflow_(wf_fullname, cust_id, cor_dic, [])

run(host='localhost', port=8181)


tasks/wf_tasks.py
from random import randrange
from sys import stdin
import requests
from json import loads, dumps
from tasks import agent
import engine.msd_wf_storage as storage
from workflow.patterns.controlflow import IF_ELSE, CHOICE


__AUTHOR__ = 'MSD'


def task_exec(exec_str, funcname):
    d = {}
    exec(exec_str) in d
    return d[funcname]


def task_dynamic_task(task_name, return_var_prefix, dict_params):
    """
    call user defined task
    dict_params is something like below:
        {
            param1: { type:'value',value:'string' or number},
            param2: { type:'var', var_name:'workflow variable name'}
            ...
        }
    """
    ddt = loads(storage.get(
        "dynamic_task:{task_name}".format(task_name=task_name)))
    RVP = return_var_prefix if return_var_prefix else ''

    def l(obj, eng):
        dict_to_pass_to_service = dict()
        for _key in dict_params.keys():
            if dict_params[_key]['type'] == 'value':
                dict_to_pass_to_service[_key] = dict_params[_key]['value']
            elif dict_params[_key]['type'] == 'var':
                dict_to_pass_to_service[_key] = obj[
                    dict_params[_key]['var_name']]

        if ddt['task_type'] == 'async':
            dict_to_pass_to_service['__callback'] = '/wf/async_task_result/{wf_name}/{cust_id}/{prefix}'.format(
                wf_name=eng._process_name, cust_id=eng.getVar('__cust_id'), prefix=RVP)

        respo = requests.post(ddt['url'], data=dumps(dict_to_pass_to_service), headers={
                              'Content-type': 'application/json'})
        respo_dict = _dynamic_task_parse_server_response(respo)
        respo_dict = {"{RVP}_{KEY}".format(RVP=RVP, KEY=key): value for (
            key, value) in respo_dict.iteritems()}
        obj.update(respo_dict)

        if ddt['task_type'] == 'async':
            eng.haltProcessing()

    return l


def _dynamic_task_parse_server_response(server_resp):
    rtn_dict = dict()
    assert isinstance(server_resp, requests.Response)
    rtn_dict["status_code"] = server_resp.status_code
    if server_resp.status_code > 299 or server_resp.status_code < 200:
        rtn_dict["body"] = server_resp.content
    else:
        if 'application/json' not in server_resp.headers['content-type'].lower().strip().split(';'):
            rtn_dict['body'] = server_resp.content
        else:
            # server response is json
            rtn_dict.update(server_resp.json())

    return rtn_dict


def assign_value_to_var(varname, value):
    def l(obj, eng):
        obj[varname] = value
    return l


def assign_var_to_var(varname1, varname2):
    def l(obj, eng):
        obj[varname1] = obj[varname2]
    return l


def test_task_add_two_number(a, b, result_var_name):
    def l(obj, eng):
        obj[result_var_name] = a + b
    return l


# ---------- TASK DEFINITIONS

def send_sms(message):
    def l(obj, eng):
        print message, " sent to ", eng.getVar('__cust_id')

    return l


def sleep_for_seconds(seconds):
    def l(obj, eng):
        print "[SYSMSG]run this process after ", seconds, "(s),again and fill input fields by these values:"
        print "User Mobile Number:", eng.getVar('__cust_id')
        print "Workflow Name:", eng._process_name
        agent.celery_execute_workflow_.apply_async(
            (eng._process_name, eng.getVar('__cust_id')), countdown=seconds)
        eng.haltProcessing()

    return l


def select_random_english_word(obj, eng):
    words = dict(cat='pishih', water='Su', nothing='Hava , Hech',
                 father='dada', mother='ana')
    m = randrange(0, len(words.keys()))
    word = words.keys()[m]
    mean = words[word]
    obj['word'] = word
    obj['mean'] = mean


def send_format_message(formatstr, variablenamesarray):
    def l(obj, eng):
        vars = dict()
        for i in variablenamesarray:
            vars[i] = obj[i]
        print formatstr.format(**vars), " sent to ", eng.getVar('__cust_id')

    return l


def get_user_guess(obj, eng):
    obj['user_guess'] = stdin.readline().strip()


def task_select_random_english_word():
    return select_random_english_word


def task_get_user_guess():
    return get_user_guess

#---------------------------------------


engine/msd_workflow_interface.py
__author__ = 'msd'
from workflow.patterns.controlflow import *

from msdworkflow import msd_workflow
from tasks.wf_tasks import *
import msd_wf_storage as storage
import json

our_workflows = {'eng_lang': [
    send_sms('Hello Dear Friend'),
    send_sms('This is Only Test'),
    sleep_for_seconds(12),
    send_sms('hi, im waked up ...'),
    send_sms('lets go to learn english words'),
    select_random_english_word,
    send_format_message(
        'do u know meaning of {word}? it is {mean}', ['word', 'mean']),
    sleep_for_seconds(3600),
    send_format_message('Oh! What is {word} meaning?', ['word']),
    get_user_guess,
    IF_ELSE(lambda obj, eng: obj['user_guess'] == obj['mean'],
            send_sms('congratulation'),
            send_sms('it was wrong. finish')
            )

]}


def get_workflow_status(wfname, wfversion, custid):
    wf_name = wfname + "." + wfversion
    key_name = get_key_name(wf_name, custid)
    if not storage.exists(key_name):
        raise wf_workflow_instance_not_exists()
    wf_instance = storage.get(key_name)
    wf = msd_workflow.deserialize_workflow(wf_instance, get_workflow_by_name)
    return dict(index=wf._i, variables=wf._objects[0])


def _compile_taskcalls(arr):
    final_array = []
    for i in arr:
        if isinstance(i, dict):
            if 'type' in i and i['type'] == 'taskcall':
                params = _compile_taskcalls(i['params'])
                method = globals()[i['taskname']]
                res = method(*params)
                final_array.append(res)
            elif 'type' in i and i['type'] == 'eval':
                final_array.append(eval(i['params'][0]))
            else:
                final_array.append(i)

        elif isinstance(i, list):
            final_array.append(_compile_taskcalls(i))

        else:
            final_array.append(i)
    return final_array


def compile_dict_to_workflow(steps_array):
    steps = _compile_taskcalls(steps_array)
    return steps
    # wf = dict(name='eng_lang',
    #           steps = [
    #               dict(type='taskcall',taskname='send_sms',params=['Hello Dear Friend']),
    #               dict(type='taskcall',taskname='send_sms',params=['This is Only Test']),
    #               dict(type='taskcall',taskname='sleep_for_seconds',params=[12]),
    #               dict(type="taskcall",taskname="send_sms",params=['hi, im waked up ...']),
    #               dict(type='taskcall',taskname="send_sms",params=['lets go to learn english words']),
    #               dict(type='taskcall',taskname='task_select_random_english_word',params=[]),
    #               dict(type='taskcall',taskname='send_format_message',params=['do u know meaning of {word}? it is {mean}', ['word', 'mean']]),
    #               dict(type='taskcall',taskname='sleep_for_seconds',params=[26]),
    #               dict(type='taskcall',taskname='send_format_message',params=['Oh! What is {word} meaning?', ['word']]),
    #               dict(type='taskcall',taskname='task_get_user_guess',params=[]),
    #               dict(type='taskcall',taskname='IF_ELSE',params=[dict(type='eval',params=["""lambda obj, eng: obj['user_guess'] == obj['mean']"""]),
    #                                                               dict(type='taskcall',taskname='send_sms',params=['congratulation']),
    #                                                               dict(type='taskcall',taskname='send_sms',params=['it was wrong. finish'])])
    #           ])


def run_workflow_(wf_fullname, cust_id, custom_vars, rtn_var_names):
    '''

    :param wf_fullname: string -- workflow fullname which maked by workflow name and its version
    :param cust_id: string -- customer id which this workflow runs for. (in new workflow)
    :param custom_vars: dict -- dictionary specifies variables in workflow in the case of new workflow they are
                            initialize variables and in the case of existing workflow it overwrites exist variables
                            in the workflow
    :param rtn_var_names: array of string -- array of varaible name which at the end of workflow execution should return.
    :return: dict -- dictionary of variables and their values specified in rtn_var_names
    '''

    wf_instace_name = get_key_name(wf_fullname, cust_id)
    if not storage.exists(wf_instace_name):
        # user in progress continue its processing
        return run_new_wf(wf_fullname, cust_id, custom_vars, rtn_var_names)
    else:
        # user is new
        return continue_wf(wf_fullname, cust_id, custom_vars, rtn_var_names)


def run_workflow(wf_name, wf_version, cust_id, custom_vars, rtn_var_names):
    '''

    :param wf_name: string -- workflow name
    :param wf_version: string -- workflow version
    :param cust_id: string -- cutomer id which this workflow runs for
    :param custom_vars: dict -- dictionary specifies variables in workflow in the case of new workflow they are
                            initialize variables and in the case of existing workflow it overwrites exist variables
                            in the workflow
    :param rtn_var_names: array of string -- array of varaible name which at the end of workflow execution should return.
    :return: dict -- dictionary of variables and their values specified in rtn_var_names
    '''
    wf_fullname = wf_name + '.' + wf_version
    return run_workflow_(wf_fullname, cust_id, custom_vars, rtn_var_names)


def get_workflow_by_name(wf_name):
    if storage.exists(wf_name):
        js = json.loads(storage.get(wf_name))
        return compile_dict_to_workflow(js)
    else:
        return []


def get_key_name(wf_name, custid):
    name = "{wf}.{custid}".format(wf=wf_name, custid=custid)
    print name
    return name


class wf_workflow_instance_not_exists(Exception):
    pass


def continue_wf(wf_name, cust_id, custom_vars, rtn_var_names):
    '''

    :param wf_name: string -- specifies full name of workflow to continue
    :param cust_id: string -- customer id who this workflow runs for
    :param custom_vars: dict -- dictionary of variable names to overwrite current variables in workflow
    :param rtn_var_names: array of string -- array of variable names which should return at the end of workflow run
    :return: return dictionary of variables and their values which specified in rtn_var_names
    '''
    key_name = get_key_name(wf_name, cust_id)
    if not storage.exists(key_name):
        raise wf_workflow_instance_not_exists()
    wf_instance = storage.get(key_name)
    wf = msd_workflow.deserialize_workflow(wf_instance, get_workflow_by_name)
    if(custom_vars is not None and isinstance(wf._objects[0], dict) and isinstance(custom_vars, dict)):
        wf._objects[0].update(custom_vars)

    is_finished = wf.continue_process()
    after_run_wf(wf, is_finished)
    return wf.get_variables(rtn_var_names)


def run_new_wf(wf_name, usermob, init_obj, rtn_var_names):
    '''

    :param wf_name: string -- specifies workflow name to run
    :param usermob: string -- customer id which this workflow runs for
    :param init_obj: dict -- dictionary of variables to initialize variables of workflow with
    :param rtn_var_names: array of string -- the variables should return after execution of workflow
    :return: dict -- return variable names specified in rtn_var_names
    '''
    wf = msd_workflow(wf_name, get_workflow_by_name)
    wf.setVar('__cust_id', usermob)
    if(init_obj is None):
        init_obj = {}
    is_finished = wf.start(init_obj)
    after_run_wf(wf, is_finished)
    return wf.get_variables(rtn_var_names)


def after_run_wf(wf, is_finished):
    wf_instace_key = get_key_name(wf._process_name, wf.getVar(
        '__cust_id'))  # wf._process_name+'_'+wf.getVar('usermob')
    if not is_finished:
        # save wf and to further run
        tmp = wf.serialize_workflow()
        storage.store(wf_instace_key, tmp)
    else:
        # delete file
        storage.remove(wf_instace_key)


if __name__ == '__main__':
    k = compile_dict_to_workflow([])
    our_workflows['eng_lang'] = k
    run_workflow('eng_lang', 'test')
